pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/model/Agent.groovy

```diff
@@ -40,47 +43,39 @@ import org.jenkinsci.plugins.scriptsecurity.sandbox.whitelists.Whitelisted
 @SuppressFBWarnings(value="SE_NO_SERIALVERSIONID")
 public class Agent implements Serializable {
     @Whitelisted
-    String docker
-
-    @Whitelisted
-    String label
-
-    @Whitelisted
-    String dockerArgs
-
-    @Whitelisted
-    Boolean any
+    Map<String,String> arguments = [:]
 
     @Whitelisted
     public Agent(Map<String,String> args) {
-        if (args.containsKey("docker")) {
-            docker = args.get("docker")
-        }
-        if (args.containsKey("label")) {
-            label = args.get("label")
-        }
-        if (args.containsKey("dockerArgs")) {
-            this.dockerArgs = args.get("dockerArgs")
-        }
+        this.arguments.putAll(args)
     }
 
+    /**
+     * Special constructor for the no-additional-arguments agent types, i.e., none and any
+     */
     @Whitelisted
-    public Agent(boolean b) {
-        this.any = b
+    public Agent(String s) {
+        this.arguments.put(s, "true")
     }
 
+    /**
+     * Get the appropriate instantiated {@link DeclarativeAgent} corresponding to our arguments.
+     *
+     * @return The instantiated declarative agent or null if not found.
+     */
     @Whitelisted
-    public boolean hasAgent() {
-        return any || docker != null || label != null
+    public DeclarativeAgent getDeclarativeAgent() {
+        String foundType = DeclarativeAgentDescriptor.orderedNames.find { arguments.containsKey(it) }
+        if (foundType != null) {
+            return DeclarativeAgentDescriptor.instanceForName(foundType, arguments)
```

Comment:
Hrm, actually, doesn't quite work - I need to find the first descriptor (ordered by ordinal-descending) with its name in the arguments list, not just a `byName`. I'll see if there's something I can do...